### ISSUE TITLE
autoplay permission

### ISSUE THREAD

**Issue by cpearce** [w3c/permissions#172]
Some web authors want to figure out whether autoplay of audible audio/video is blocked by the UA, so they can avoid downloading the video unless they're allowed to play it. Maybe the Permissions spec is a reasonable place for us to expose that? Then web authors don't have to jump through hoops to determine whether they can play?

For example, [we had a bug filed against Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1442186) because it appears that our block autoplay implementation wasn't working with their script to detect whether autoplay worked.

**Comment by mounirlamouri** [comment]
It's a use case I've heard too in the past and after talking to Apple, it seems that we could also get Safari on board. However, I'm not sure if a permission is the right approach as autoplay abilities may change during the lifetime of the document/element depending on the autoplay policy implemented by the browser. For example, with the new Chrome autoplay policy, most documents will start without autoplay abilities and will most likely get it before being closed. Technically, this shouldn't prevent using a permission but the concept seems off as it's not always linked to the origin and might not be persistent.

I think we may want to add some mechanism to the Document or the HTMLMediaElement to expose the autoplay ability, depending on what browsers can agree on.

**Comment by cpearce** [comment]
I don't have a deep knowledge of the permissions spec, but I wanted to clarify this...

> autoplay abilities may change during the lifetime of the document/element depending on the autoplay policy implemented by the browser.

Don't other permissions change during the lifetime of a document though? For example, after the UA prompts for camera/mic access, and the user grants permission, the observable permission would change there?

> Technically, this shouldn't prevent using a permission but the concept seems off as it's not always linked to the origin and might not be persistent.

Isn't camera/mic access tied to an origin though?

**Comment by mounirlamouri** [comment]
>Don't other permissions change during the lifetime of a document though? For example, after the UA prompts for camera/mic access, and the user grants permission, the observable permission would change there?

That's true. And to clarify, what I meant is that it would be in a binary state `denied` or `granted` and will, most likely start as `denied` then switch to `granted` on every visit. This isn't a common pattern for most permissions. When using the Permissions API, it's fair to assume that a `deny` permission will not be easily overturned.

>>Technically, this shouldn't prevent using a permission but the concept seems off as it's not always linked to the origin and might not be persistent.

> Isn't camera/mic access tied to an origin though?

My point was that permissions are linked to origins but autoplay may or may not. I don't know of a permission that isn't origin-bound at least in Chromium.

**Comment by cpearce** [comment]
Maybe it's better to have a member on HTMLMediaElement, such as:
`readonly bool canAutoplay;` or `promise<bool> canAutoplay();`.
that takes into account the current context. We should probably move this request to the WhatWG spec.

**Comment by travisleithead** [comment]
Hey, so cool to see this issue! We were just discussing this over in [the TAG's design-review repo](https://github.com/w3ctag/design-reviews/issues/203) and this existing issue saves me from opening a new one to also convey our feedback 😊

We [the TAG] also agree that this is an important need. After deliberating, our conclusion seemed to land with modelling this as a Permission too. Here's the official resolution:

> Media autoplay should be under user control, that could mean for example having a setting "yes, autoplay is OK across the board" or it could mean origin-specific permissions prompts. "yes, netflix can autoplay videos". ---> therefore modeling autoplay as a permission is desirable because the diversity of effective policies will require that sites be able to introspect the policy without attempting to play media.

We're happy to discuss further! Thanks for indulging my interjection here.

**Comment by mounirlamouri** [comment]
@cpearce I've opened https://github.com/whatwg/html/issues/3617 I think we should be able to craft a solution over there.

@travisleithead I do not believe a permission is going to be of any help for web developers given that the autoplay state of a media element varies based on the user actions on the page and the state of the media element. I do not believe that an autoplay permission would be of any help to users too as autoplay is a under-defined concept that web developers, browser developers and users all perceive differently. Chrome decided to not expose such permission/setting because we believe we will not be able to genuinely "block" autoplay on a website to the standard users expect.

Closing so we can move the discussion to the WHATWG issue.
